File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.cpp
#include <string>
#include <vector>
#include <unordered_set>
#include <algorithm>

using namespace std;

bool solution(vector<string> phone_book) {
    unordered_set<string> phone_numbers;
    sort(phone_book.begin(), phone_book.end());
    
    for (int i = 0; i < phone_book.size(); i++) {
        if (phone_numbers.empty()) {
            phone_numbers.insert(phone_book[i]);
        }
        else {
            string substr = "";
            for (int j = 0; j < phone_book[i].length(); j++) {
                substr += phone_book[i][j];
                if (phone_numbers.find(substr) != phone_numbers.end())
                    return false;
            }
            phone_numbers.insert(phone_book[i]);
        }
    }
    
    return true;
}

File: yjKim513/프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    phone_book = sorted(phone_book)
    
    i = 0
    while i < len(phone_book) - 1:
        if phone_book[i] == phone_book[i + 1][:len(phone_book[i])]:
            return False
        i += 1
    
    return True